File: src/core/state.ts
/**
 * 本地同步状态 `~/.homer/state.json`（docs/m2-plan.md §2.1，P0 完整实现）。
 *
 * 冻结签名：
 *   HomerState / loadState(paths): HomerState / saveState(paths, state): void
 *
 * 语义（计划原文）：
 *   - `loadState`：文件缺失 / 损坏 → `{ version: 1 }`，**不 throw**（同步状态是加速信息，
 *     任何读取失败都必须降级为「无 base」而不是让 CLI 崩掉；对应 §1 决策 D3 的 fallback 链）。
 *   - `saveState`：tmp + rename 原子写（同 config.saveConfig），避免写到一半崩溃留下坏 state
 *     被下一轮读成 `lastSyncCommit` 缺失 / 截断的脏值。
 *
 * state.json 与 `backups/` 一样**不入库**（§1 决策 D5）：同步策略本身不参与同步。
 */

import fs from 'node:fs';
import path from 'node:path';

import type { HomerPaths } from './paths.js';
import { isPlainObject } from './entry-kind.js';

/** `lastSyncCommand` 的取值集合（冻结，与 sync/state 记录口径一致）。 */
const SYNC_COMMANDS = ['push', 'pull', 'merge'] as const;

export interface HomerState {
  version: 1;
  lastSyncCommit?: string;                      // 上次同步成功后的 HEAD SHA
  lastSyncAt?: string;                          // ISO 8601
  lastSyncCommand?: 'push' | 'pull' | 'merge';
}

/** 空状态（文件缺失 / 损坏时的降级值）。 */
function emptyState(): HomerState {
  return { version: 1 };
}

/**
 * 读取 state.json。
 * 任何失败（ENOENT / 非 JSON / 顶层非对象）→ `{ version: 1 }`。
 * 字段级容忍：类型不对的可选字段直接丢弃（不抛、不猜），未知键忽略。
 */
export function loadState(paths: HomerPaths): HomerState {
  let text: string;
  try {
    text = fs.readFileSync(paths.stateFile, 'utf8');
  } catch {
    return emptyState();
  }

  let raw: unknown;
  try {
    raw = JSON.parse(text) as unknown;
  } catch {
    return emptyState();
  }

  if (!isPlainObject(raw)) return emptyState();

  const state = emptyState();

  const commit = raw['lastSyncCommit'];
  if (typeof commit === 'string' && commit.trim() !== '') state.lastSyncCommit = commit;

  const at = raw['lastSyncAt'];
  if (typeof at === 'string' && at.trim() !== '') state.lastSyncAt = at;

  const command = raw['lastSyncCommand'];
  if (typeof command === 'string' && (SYNC_COMMANDS as readonly string[]).includes(command)) {
    state.lastSyncCommand = command as HomerState['lastSyncCommand'];
  }

  return state;
}

/**
 * 写入 state.json（2 空格缩进 + 末尾换行，与 homer.json 手写格式一致）。
 * 先写 `<stateFile>.tmp-<pid>` 再 rename：同目录 rename 在同一文件系统上是原子的，
 * 读侧要么看到旧版完整内容、要么看到新版完整内容。
 */
export function saveState(paths: HomerPaths, state: HomerState): void {
  fs.mkdirSync(path.dirname(paths.stateFile), { recursive: true });
  const tmp = `${paths.stateFile}.tmp-${process.pid}`;
  fs.writeFileSync(tmp, `${JSON.stringify(state, null, 2)}\n`, 'utf8');
  fs.renameSync(tmp, paths.stateFile);
}

File: internal/core/state.go
package core

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/zzjcool/homer-cli/internal/orderedjson"
)

// PairedDevice records the public identity of a machine that has completed
// an online key pairing. It is metadata only; the private identity remains
// outside the repository.
type PairedDevice struct {
	Hostname  string `json:"hostname"`
	Recipient string `json:"recipient"`
	PairedAt  string `json:"pairedAt"`
}

// HomerState records the last successfully synchronized repository revision.
// It is acceleration metadata, not synchronization content.
type HomerState struct {
	Version         int            `json:"version"`
	LastSyncCommit  string         `json:"lastSyncCommit,omitempty"`
	LastSyncAt      string         `json:"lastSyncAt,omitempty"`
	LastSyncCommand string         `json:"lastSyncCommand,omitempty"`
	Paired          []PairedDevice `json:"paired,omitempty"`
}

// AddPairedDevice returns a copy with device merged into the recorded list.
// Recipient is the stable deduplication key: updating a known recipient
// preserves its position while replacing its hostname and pairing time.
func (s HomerState) AddPairedDevice(device PairedDevice) HomerState {
	merged := s
	merged.Paired = append([]PairedDevice(nil), s.Paired...)
	for index := range merged.Paired {
		if merged.Paired[index].Recipient != device.Recipient {
			continue
		}
		merged.Paired[index].Hostname = device.Hostname
		merged.Paired[index].PairedAt = device.PairedAt
		return merged
	}
	merged.Paired = append(merged.Paired, device)
	return merged
}

// PairedDisplay renders the deliberately abbreviated recipient form used in
// human-facing reports. State is non-authoritative, so malformed entries are
// ignored rather than allowed to produce an unsafe or misleading line.
func PairedDisplay(state HomerState) []string {
	displays := make([]string, 0, len(state.Paired))
	for _, device := range state.Paired {
		if !pairedDeviceDisplayValid(device) {
			continue
		}
		recipient := device.Recipient
		fingerprint := recipient[:12] + "…" + recipient[len(recipient)-4:]
		displays = append(displays, device.Hostname+"（"+fingerprint+"）")
	}
	if len(displays) == 0 {
		return nil
	}
	return displays
}

func pairedDeviceDisplayValid(device PairedDevice) bool {
	if strings.TrimSpace(device.Hostname) == "" || strings.TrimSpace(device.Recipient) == "" || strings.TrimSpace(device.PairedAt) == "" {
		return false
	}
	// The display needs enough bytes for the frozen first-12/last-4
	// fingerprint. Recipients are ASCII, so byte slicing is intentional.
	return len(device.Recipient) > 16
}

func emptyState() HomerState { return HomerState{Version: 1} }

func validSyncCommand(command string) bool {
	return command == "push" || command == "pull" || command == "merge"
}

// LoadState treats every read, parse, and shape failure as an empty state.
// State is deliberately non-authoritative: callers can fall back to the
// workspace store when this acceleration file is absent or damaged.
func LoadState(paths HomerPaths) HomerState {
	data, err := os.ReadFile(paths.StateFile)
	if err != nil {
		return emptyState()
	}

	parsed, err := orderedjson.Parse(data)
	if err != nil {
		return emptyState()
	}
	object, ok := parsed.(*orderedjson.Object)
	if !ok || object == nil {
		return emptyState()
	}

	state := emptyState()
	if commit, ok := object.M["lastSyncCommit"].(string); ok && strings.TrimSpace(commit) != "" {
		state.LastSyncCommit = commit
	}
	if at, ok := object.M["lastSyncAt"].(string); ok && strings.TrimSpace(at) != "" {
		state.LastSyncAt = at
	}
	if command, ok := object.M["lastSyncCommand"].(string); ok && validSyncCommand(command) {
		state.LastSyncCommand = command
	}
	if paired, ok := parsePaired(object.M["paired"]); ok {
		state.Paired = paired
	}
	return state
}

// parsePaired applies the deliberately narrow shape check for the additive
// paired field. A single malformed element invalidates the whole field, but
// never invalidates the older last-sync metadata in the same state object.
func parsePaired(value orderedjson.Value) ([]PairedDevice, bool) {
	items, ok := value.([]orderedjson.Value)
	if !ok {
		return nil, false
	}
	paired := make([]PairedDevice, 0, len(items))
	for _, item := range items {
		object, ok := item.(*orderedjson.Object)
		if !ok || object == nil {
			return nil, false
		}
		hostname, hostnameOK := object.M["hostname"].(string)
		recipient, recipientOK := object.M["recipient"].(string)
		pairedAt, pairedAtOK := object.M["pairedAt"].(string)
		if !hostnameOK || !recipientOK || !pairedAtOK {
			return nil, false
		}
		paired = append(paired, PairedDevice{
			Hostname:  hostname,
			Recipient: recipient,
			PairedAt:  pairedAt,
		})
	}
	return paired, true
}

func loadState(paths HomerPaths) HomerState { return LoadState(paths) }

func stateValue(state HomerState) orderedjson.Value {
	keys := []string{"version"}
	values := map[string]orderedjson.Value{
		"version": orderedStateNumber(state.Version),
	}
	if state.LastSyncCommit != "" {
		keys = append(keys, "lastSyncCommit")
		values["lastSyncCommit"] = state.LastSyncCommit
	}
	if state.LastSyncAt != "" {
		keys = append(keys, "lastSyncAt")
		values["lastSyncAt"] = state.LastSyncAt
	}
	if state.LastSyncCommand != "" {
		keys = append(keys, "lastSyncCommand")
		values["lastSyncCommand"] = state.LastSyncCommand
	}
	if len(state.Paired) > 0 {
		keys = append(keys, "paired")
		items := make([]orderedjson.Value, 0, len(state.Paired))
		for _, device := range state.Paired {
			items = append(items, &orderedjson.Object{
				Keys: []string{"hostname", "recipient", "pairedAt"},
				M: map[string]orderedjson.Value{
					"hostname":  device.Hostname,
					"recipient": device.Recipient,
					"pairedAt":  device.PairedAt,
				},
			})
		}
		values["paired"] = items
	}
	return &orderedjson.Object{Keys: keys, M: values}
}

func orderedStateNumber(value int) orderedjson.Value {
	parsed, err := orderedjson.Parse([]byte(fmt.Sprint(value)))
	if err != nil {
		return nil
	}
	return parsed
}

// SaveState atomically replaces state.json. The temporary file is in the same
// directory, so rename has the same-filesystem atomicity required by the
// frozen state contract.
func SaveState(paths HomerPaths, state HomerState) error {
	if err := os.MkdirAll(filepath.Dir(paths.StateFile), 0o777); err != nil {
		return err
	}
	if state.Version == 0 {
		state.Version = 1
	}

	data := orderedjson.SerializeFile(stateValue(state))
	tmp := fmt.Sprintf("%s.tmp-%d", paths.StateFile, os.Getpid())
	if err := os.WriteFile(tmp, data, 0o666); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	if err := os.Rename(tmp, paths.StateFile); err != nil {
		_ = os.Remove(tmp)
		return err
	}
	return nil
}

func saveState(paths HomerPaths, state HomerState) error { return SaveState(paths, state) }

// StateValue exposes the ordered representation for git/diagnostic workers
// without introducing a second JSON serialization implementation.
func StateValue(state HomerState) orderedjson.Value { return stateValue(state) }
